engines/prover.cpp

```diff
@@ -68,10 +70,284 @@ void Prover::initialize()
 {
   reached_k_ = -1;
   bad_ = solver_->make_term(smt::PrimOp::Not, property_.prop());
+  if (options_.static_coi_) {
+    if (!ts_.is_functional())
+      throw PonoException("Temporary restriction: cone-of-influence analysis currently "\
+                          "supported for functional transition systems only.");
+    /* Compute the set of state/input variables related to the
+       bad-state property. Based on that information, rebuild the
+       transition relation of the transition system. */
+    compute_coi();
+    orig_num_statevars_ = ts_.statevars().size();
+    orig_num_inputvars_ = ts_.inputvars().size();
+    ts_.rebuild_trans_based_on_coi(statevars_in_coi_, inputvars_in_coi_);
+    assert(statevars_in_coi_.size() == ts_.statevars().size());
+    assert(inputvars_in_coi_.size() == ts_.inputvars().size());
+    logger.log(
+        1,
+        "COI analysis completed: {} remaining input variables, {} original",
+        inputvars_in_coi_.size(),
+        orig_num_inputvars_);
+    logger.log(
+        1,
+        "COI analysis completed: {} remaining state variables, {} original",
+        statevars_in_coi_.size(),
+        orig_num_statevars_);
+  }
 }
 
 ProverResult Prover::prove() { return check_until(INT_MAX); }
 
+/* For debugging only. */
+void Prover::print_term_dfs(const Term & term)
+{
+  UnorderedTermSet visited_terms;
+  TermVec open_terms;
+  open_terms.push_back (term);
+  Term cur;
+
+  while (!open_terms.empty()) {
+    cur = open_terms.back();
+    open_terms.pop_back ();
+
+    if (visited_terms.find(cur) == visited_terms.end()) {
+      // cache 'cur' and push its children
+      visited_terms.insert(cur);
+
+      cout << "  visiting term: " << cur << "\n";
+      if (cur->is_symbol())
+        cout << "    ..is symbol\n";
+      
+      for (auto child : cur) {
+        cout << "    pushing child: " << child << "\n";
+        open_terms.push_back(child);
+      }
+    }
+  }
+}
+
+/* For debugging only. */
+void Prover::print_coi_info()
+{
+  cout << "TEST PRINT COI\n";
+  cout << "bad_ term: " << bad_ << "\n";
+  print_term_dfs(bad_);
+
+  cout << "init_ term: " << ts_.init() << "\n";
+  print_term_dfs(ts_.init());
+
+  cout << "trans_ term: " << ts_.trans() << "\n";
+  print_term_dfs(ts_.trans());
+
+  cout << "input vars: \n";
+  for (auto inputvar : ts_.inputvars())
+    cout << "  " << inputvar << "\n";
+
+  cout << "state vars: \n";
+  for (auto statevar : ts_.statevars())
+    cout << "  " << statevar << "\n";
+
+  cout << "constraints: \n";
+  for (auto constr : ts_.constraints())
+    cout << "  " << constr << "\n";  
+}
+
+/* Add 'term' to 'set' if it does not already appear there. */
+void Prover::collect_coi_term(UnorderedTermSet & set, const Term & term)
```

Comment:
this could be a static non-member function

Reply:
Like above, I agree. I suggest to leave it for now as it will change during refactoring.